File: Shared/Obj-c/ExceptionBridging/ExceptionBridging.m
// Copyright © 2022 Tokenary. All rights reserved.

#import "ExceptionBridging.h"

NS_ASSUME_NONNULL_BEGIN

@implementation NSObject(ObjC_Exception)

+ (void) objc_try:(void (^) (void)) tryBlock
       objc_catch:(void (^) (NSException* exception)) catchBlock
       objc_final:(nullable void (^) (void)) finallyBlock
{
    @try {
        tryBlock();
    } @catch (NSException* exception) {
        catchBlock(exception);
    } @finally {
        finallyBlock();
    }
}

+ (void) objc_throw:(NSException*) exception {
    @throw exception;
}

@end

NS_INLINE void objc_try(
                        void (^ tryBlock)(void),
                        void (^ catchBlock)(NSException* exception),
                        void (^_Nullable finallyBlock)(void)
) {
    @try {
        tryBlock();
    } @catch (NSException* exception) {
        catchBlock(exception);
    } @finally {
        finallyBlock();
    }
}

NS_INLINE void objc_throw(NSException* exception) {
    @throw exception;
}

NS_ASSUME_NONNULL_END

File: Shared/Obj-c/ExceptionBridging/ExceptionBridging.swift
// Copyright © 2022 Tokenary. All rights reserved.

import Foundation

public func objc(
    try tryBlock: @autoclosure @escaping () -> Void,
    catch catchBlock: @escaping (_ exception: NSException) -> Void,
    finally finallyBlock: (() -> Void)? = nil
) {
    NSObject.objc_try(tryBlock, objc_catch: catchBlock, objc_final: finallyBlock)
}

public func objc(
    throw objCException: (name: String, message: String?, userInfo: [AnyHashable: Any]?)
) {
    NSObject.objc_throw(
        NSException(
            name: NSExceptionName(objCException.name),
            reason: objCException.message ?? objCException.name,
            userInfo: objCException.userInfo
        )
    )
}
